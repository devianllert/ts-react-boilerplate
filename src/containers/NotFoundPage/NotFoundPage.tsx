import React, { ReactElement } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';

import Button from '../../components/Button';

import styles from './NotFoundPage.module.scss';

const NotFoundPage = (): ReactElement => {
  const { t } = useTranslation();

  return (
    <div className={styles.content}>
      <Helmet>
        <title>{t('notFound.headTitle')}</title>
      </Helmet>

      <h1 className={styles.logo}>404</h1>
      <h2 className={styles.title}>{t('notFound.title')}</h2>
      <span className={styles.subtitle}>{t('notFound.subtitle')}</span>

      <Button to="/">{t('notFound.button')}</Button>
    </div>
  );
};

export default NotFoundPage;
