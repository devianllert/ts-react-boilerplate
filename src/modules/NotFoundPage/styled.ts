import styled from 'styled-components';

export const Image = styled.img`
  width: 100%;
  max-height: 400px;
  margin-bottom: 16px;
`;

export const CenteredBox = styled.div`
  min-height: 100vh;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;
