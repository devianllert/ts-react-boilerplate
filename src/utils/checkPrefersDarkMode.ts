/**
 * Check browser or OS preference for dark mode
 */

const checkPrefersDarkMode = (): boolean => {
  const preferDarkQuery = '(prefers-color-scheme: dark)';

  const mediaQuery = matchMedia(preferDarkQuery);

  return mediaQuery.media === preferDarkQuery;
};

export default checkPrefersDarkMode;
