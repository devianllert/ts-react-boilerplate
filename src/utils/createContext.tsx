import React, {
  createContext,
  useState,
  useContext,
  Provider,
  Dispatch,
  SetStateAction,
  PropsWithChildren,
  ReactNode,
  Context,
} from 'react';

type UpdateType<A> = Dispatch<SetStateAction<A>>;

/**
 * A helper to create a Context and Provider with no upfront default value, and
 * without having to check for undefined all the time.
 */
export const createCtx = <A extends {} | null>(): readonly [() => A, Provider<A | undefined>] => {
  const ctx = createContext<A | undefined>(undefined);

  const useCtx = (): A => {
    const c = useContext(ctx);

    if (c === undefined) {
      throw new Error('useCtx must be inside a Provider with a value');
    }

    return c;
  };

  return [useCtx, ctx.Provider] as const;
};

export const createCtxWithUpdate = <A extends any>(
  defaultValue: A,
): readonly [
  Context<{ state: A; update: Dispatch<SetStateAction<A>> }>,
  (props: { children?: ReactNode }) => ReactNode,
] => {
  const defaultUpdate: UpdateType<typeof defaultValue> = () => defaultValue;

  const ctx = createContext({
    state: defaultValue,
    update: defaultUpdate,
  });

  const ContextProvider = (props: PropsWithChildren<{}>): ReactNode => {
    const [state, update] = useState(defaultValue);
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <ctx.Provider value={{ state, update }} {...props} />;
  };

  return [ctx, ContextProvider] as const;
};
