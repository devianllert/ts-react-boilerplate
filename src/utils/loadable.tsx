/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react/jsx-props-no-spreading */

import React, {
  lazy,
  Suspense,
  ReactElement,
  ComponentType,
} from 'react';

export type ImportComponent = Promise<{ default: any }>;

const loadable = (importFunc: () => ImportComponent, { fallback = null }): ComponentType => {
  const LazyComponent: ComponentType = lazy(importFunc);

  return (props: object): ReactElement => (
    <Suspense fallback={fallback}>
      <LazyComponent {...props} />
    </Suspense>
  );
};

export default loadable;
