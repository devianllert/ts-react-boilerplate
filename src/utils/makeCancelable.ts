/* eslint-disable prefer-promise-reject-errors */
/* https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html */
export interface CancellablePromise<T> {
  promise: Promise<T>;
  cancel(): void;
}

const makeCancelable = <T>(promise: Promise<T>): CancellablePromise<T> => {
  let hasCanceled = false;

  const wrappedPromise = new Promise<T>((resolve, reject) => {
    promise.then(
      (val) => (hasCanceled ? reject({ isCanceled: true }) : resolve(val)),
      (error) => (hasCanceled ? reject({ isCanceled: true }) : reject(error)),
    );
  });

  return {
    promise: wrappedPromise,
    cancel(): void {
      hasCanceled = true;
    },
  };
};

export default makeCancelable;
