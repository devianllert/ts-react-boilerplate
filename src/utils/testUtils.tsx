/* eslint-disable import/prefer-default-export */
import React, { ReactElement } from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Provider } from 'react-redux';
import { i18n as i18Next } from 'i18next';
import { I18nextProvider } from 'react-i18next';
import { render, RenderResult } from '@testing-library/react'; // eslint-disable-line

import configureStore, { EnhancedStore } from '../core/configureStore';
import i18nInstance from '../core/i18n';

export interface RenderProvidersOptions {
  initialState?: {};
  route?: string;
  routerHistory?: MemoryHistory;
  store?: EnhancedStore;
  i18n?: i18Next;
}

export interface RenderProvidersResult extends RenderResult {
  routerHistory: MemoryHistory;
  store: EnhancedStore;
  i18n: i18Next;
}

export type RenderProviders = (ui: ReactElement, options?: RenderProvidersOptions) => RenderProvidersResult;

const renderWithProviders: RenderProviders = (
  ui,
  {
    initialState = {},
    route = '/',
    routerHistory = createMemoryHistory({ initialEntries: [route] }),
    store = configureStore(initialState),
    i18n = i18nInstance,
  }: RenderProvidersOptions = {},
): RenderProvidersResult => ({
  ...render(
    <Provider store={store}>
      <I18nextProvider i18n={i18n}>
        <Router history={routerHistory}>{ui}</Router>
      </I18nextProvider>
    </Provider>,
  ),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  store,
  routerHistory,
  i18n,
});

export {
  renderWithProviders,
};
