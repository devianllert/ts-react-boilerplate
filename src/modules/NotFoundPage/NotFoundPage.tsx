import React, { ReactElement } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { MdArrowBack } from 'react-icons/md';

import Button from '../../components/Button';
import Typography from '../../components/Typography';

import * as S from './styled';

const NotFoundPage = (): ReactElement => {
  const { t } = useTranslation();

  return (
    <>
      <Helmet>
        <title>{t('NOT_FOUND_TITLE')}</title>
      </Helmet>

      <S.CenteredContainer>
        <S.Image src="/img/not_found.svg" alt="not found" />

        <Typography variant="h1" align="center" gutterBottom>404</Typography>
        <Typography variant="h4" align="center" gutterBottom>{t('NOT_FOUND_TITLE')}</Typography>

        <Button to="/" startIcon={<MdArrowBack size={16} />}>{t('BACK')}</Button>
      </S.CenteredContainer>
    </>
  );
};

export default NotFoundPage;
