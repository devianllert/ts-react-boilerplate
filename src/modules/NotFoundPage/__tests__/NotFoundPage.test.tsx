import React from 'react';
import { cleanup } from '@testing-library/react';

import { renderWithProviders } from '../../../utils/testUtils';

import NotFoundPage from '../NotFoundPage';

describe('<NotFoundPage />', (): void => {
  afterEach(cleanup);

  it('should render and match the snapshot', (): void => {
    const { container } = renderWithProviders(<NotFoundPage />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with initial state', (): void => {
    const { getByText } = renderWithProviders(<NotFoundPage />);

    expect(getByText(/404/)).toBeDefined();
  });
});
