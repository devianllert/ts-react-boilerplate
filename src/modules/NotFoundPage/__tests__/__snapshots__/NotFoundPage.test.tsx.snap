// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotFoundPage /> should render and match the snapshot 1`] = `
.c0 {
  box-sizing: border-box;
  display: block;
  width: 100%;
  margin-left: auto;
  margin-right: auto;
  padding-left: 12px;
  padding-right: 12px;
}

.c5 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  min-width: 88px;
  height: 36px;
  padding: 0 16px;
  text-align: center;
  vertical-align: middle;
  font-family: "Roboto",sans-serif;
  font-size: 14px;
  font-weight: 500;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  border: 0;
  border-radius: 4px;
  color: rgba(255,255,255,0.87);
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: uppercase;
  -webkit-tap-highlight-color: transparent;
  outline: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  touch-action: manipulation;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  -webkit-transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,box-shadow 250ms cubic-bezier(0.4,0,0.2,1) 0ms,border 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,box-shadow 250ms cubic-bezier(0.4,0,0.2,1) 0ms,border 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c5:before {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: transparent;
  -webkit-transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
}

.c5:hover:before {
  background-color: rgba(255,255,255,0.12);
}

.c5:focus:before {
  background-color: rgba(255,255,255,0.18);
}

.c5:active:before {
  background-color: rgba(255,255,255,0.24);
}

.c5:disabled {
  box-shadow: none;
}

.c6 {
  width: 100%;
  display: inherit;
  -webkit-align-items: inherit;
  -webkit-box-align: inherit;
  -ms-flex-align: inherit;
  align-items: inherit;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
}

.c7 {
  display: inherit;
  margin-right: 8px;
  margin-left: -4px;
}

.c3 {
  margin: 0;
  color: initial;
  text-align: center;
  margin-bottom: 0.35em;
  font-size: 9.6rem;
  font-weight: 300;
  -webkit-letter-spacing: -1.5px;
  -moz-letter-spacing: -1.5px;
  -ms-letter-spacing: -1.5px;
  letter-spacing: -1.5px;
  line-height: 1.167;
}

.c4 {
  margin: 0;
  color: initial;
  text-align: center;
  margin-bottom: 16px;
  font-size: 3.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0.25px;
  -moz-letter-spacing: 0.25px;
  -ms-letter-spacing: 0.25px;
  letter-spacing: 0.25px;
  line-height: 1.235;
}

.c2 {
  width: 100%;
  max-height: 400px;
  margin-bottom: 16px;
}

.c1 {
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

@media (min-width:960px) {
  .c0 {
    padding-left: 16px;
    padding-right: 16px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <svg
      class="c2"
    >
      not_found.svg
    </svg>
    <h1
      class="c3"
      color="initial"
      display="initial"
    >
      404
    </h1>
    <p
      class="c4"
      color="initial"
      display="initial"
    >
      NOT_FOUND_TITLE
    </p>
    <a
      appearence="primary"
      class="c5"
      href="/"
    >
      <span
        class="c6"
      >
        <span
          class="c7"
        >
          <svg
            fill="currentColor"
            height="16"
            size="16"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        BACK
      </span>
    </a>
  </div>
</div>
`;
